packages/react-scripts/config/paths.js

```diff
@@ -46,19 +46,44 @@ function getServedPath(appPackageJson) {
   return ensureSlash(servedUrl, true);
 }
 
+const isFlow = fs.existsSync(resolveApp('.flowconfig'));
+const isTypeScript = fs.existsSync(resolveApp('tsconfig.json'));
+const hasTSConfigProd = fs.existsSync(resolveApp('tsconfig.prod.json'));
+const useTSLint = fs.existsSync(resolveApp('tslint.json'));
+const hasTSLintProd = fs.existsSync(resolveApp('tslint.prod.json'));
+
+const resolveTS = (
+  resolveFn,
+  baseFilePath,
+  extension = 'ts',
+  checkIfExists = true
+) => {
+  const jsFilePath = resolveFn(`${baseFilePath}.js`);
+  if (!isTypeScript) {
+    return jsFilePath;
+  }
+
+  const tsFilePath = resolveFn(`${baseFilePath}.${extension}`);
+  return !checkIfExists || fs.existsSync(tsFilePath) ? tsFilePath : jsFilePath;
+};
+
 // config after eject: we're in ./config/
 module.exports = {
   dotenv: resolveApp('.env'),
   appPath: resolveApp('.'),
   appBuild: resolveApp('build'),
   appPublic: resolveApp('public'),
   appHtml: resolveApp('public/index.html'),
-  appIndexJs: resolveApp('src/index.js'),
+  appIndexJs: resolveTS(resolveApp, 'src/index', 'tsx'),
```

Comment:
Let's make a more generic resolve helper that follows the same order as webpack will. There's no use in us trying to do something tricky imo.

Reply:
@Timer are you saying something like `resolveJS('src/index')` that first tries to detect `.mjs`, then `.jsx`, `ts` and `tsx`?

Reply:
`resolveModule` is probably a better name, but yes C+P the extension order.

Reply:
@Timer ok, your suggestions are in